test/cql-pytest/test_describe.py

```diff
@@ -0,0 +1,166 @@
+# Copyright 2022-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+#############################################################################
+# Tests for server-side describe
+#############################################################################
+
+import pytest
+from util import new_test_table, new_type, new_function, new_aggregate
+
+# Extract names from listing describe result
+def extract_names(result):
+    return list(map(lambda r: r.name, result))
+
+# Extract descriptions from describe result and join them to one string
+def extract_describe(result):
+    descs = list(map(lambda r: r.create_statement, result))
+    desc = " ".join(descs)
+    return desc
+
+# Test that `DESC KEYSPACES` contains newly created user keyspace
+def test_desc_keyspaces(cql, test_keyspace):
+        result = cql.execute("DESC KEYSPACES")
+        names = extract_names(result)
+        assert test_keyspace in names
+
+# Test that `DESC KEYSPACE {name}` contains create statements for keyspace and table
+# and `DESC ONLY KEYSPACE {name}` contains only create statement for keyspace
+def test_desc_keyspace(cql, test_keyspace):
+    with new_test_table(cql, test_keyspace, "id int primary key") as table:
+        result = cql.execute(f"DESC KEYSPACE {test_keyspace}")
+        desc = extract_describe(result)
+
+        assert f"CREATE KEYSPACE {test_keyspace}" in desc
+        assert f"CREATE TABLE {table}" in desc
+
+        only_result = cql.execute(f"DESC ONLY KEYSPACE {test_keyspace}")
+        only_desc = extract_describe(only_result)
+
+        assert f"CREATE KEYSPACE {test_keyspace}" in only_desc
+        assert f"CREATE TABLE {table}" not in only_desc
+
+# Test that `DESC TABLES` contains some system table
+# and newly created table
+def test_desc_tables(cql, test_keyspace):
+    result = cql.execute("DESC TABLES")
+    names = extract_names(result)
+
+    assert "local" in names
+    assert "sstable_activity" in names
+
+    with new_test_table(cql, test_keyspace, "id int primary key") as table:
+        table_name = table.split(".")[1]
+        cql.execute(f"use {test_keyspace}")
+        ks_result = cql.execute("DESC TABLES")
+        ks_names = extract_names(ks_result)
+
+        assert "local" not in ks_names
+        assert "sstable_activity" not in ks_names
+        assert table_name in ks_names
+
+# Test that `DESC TABLE {name}` contains create statement for table
+def test_desc_table(cql, test_keyspace):
+    with new_test_table(cql, test_keyspace, "id int primary key") as table:
+        table_name = table.split(".")[1]
+        result = cql.execute(f"DESC TABLE {test_keyspace}.{table_name}")
+        desc = extract_describe(result)
+
+        assert f"CREATE TABLE {test_keyspace}.{table_name}" in desc
+        assert "id int" in desc
+        assert "PRIMARY KEY (id)" in desc
```

Comment:
I think a better test for `DESC TABLE` and maybe all the other `DESC` statements is a circular test:
* create table
* describe table (1)
* create table reusing statement supplied by the above
* describe table (2)

Check that (1) == (2) (with a different name of course)

Reply:
It's not "better" in the sense that:
1. If "DESC TABLE" always returned an empty string (or some other silly statement), your test will pass :-)
2. If "DESC TABLE" is missing one specific attribute, your test will pass also.
I agree it's better if somehow our output is very different from Cassandra's but both are correct, so it's hard to write a test which passes on both. But is this the case? Would would our output be very different from Cassandra's?

Reply:
We could go further and read `system_schema.*` tables to make sure the table created by feeding back the result of the `DESC` statement yields a table identical to the original (again, except the name).

Reply:
This is a nice idea. It assumes that the old pre-server-side-describe code took all the information from system_schema, so if the information there is identical, then so are the table.

The only problem with this is that in the future, we or Cassandra may invent new table parameters and not bother to put them in system_schema (because the old cqlsh doesn't understand them anyway), so wrong output for these table parameters cannot be seen in system_schema. I don't know how likely this problem is, however.

The good thing about your suggestion is that it makes it very easy to test 50 different table-creation commands in one loop without a lot of repetitions in the code.

Reply:
Yes, the idea is to have one large randomized test covering a very wide aspect of the tested code. Then small tests can be added for specific corner cases.
My usual problem when testing things related to schema is that developers tend to come up with very simple schemas and don't tend to use most/any of the table options. This ends up with poor coverage of the schema-related logic.
We need to generate random schemas for this test, which:
* can generate [1, max_PK) partition key columns with every possible type (including frozen collections, UDTs, etc)
* can generate [0, max_CK) clustering key columns with every possible type
* can generate [1, max_regular] regular columns with every possible type
* can generate [0, max_static] static columns with every possible type
* can generate every legal combination of the table options
* can generate counter tables (these usually get less attention)

The idea is not to create a table with every possible column with every possible type, but to have a generating code that is able to generate every possible combination. Then in time, through all the CI, next promotion and developer runs this creates a very good coverage.
Care must be taken that the schemas generated have structural variety as well: we want to generate schemas with all features (clustering columns, static columns, etc), but we also want to generate much simpler schemas too, e.g. a key-value schema.